### PR TITLE
Added configurable strict loading for Fabric strategies

### LINKED ISSUE #16447: Make strict loading configurable in `Fabric.load()`
### Description & Motivation

We added a new API for loading in Fabric in https://github.com/Lightning-AI/lightning/pull/16434. The loading of the module parameters currently uses just `strict=True`. 

### Pitch

Make it an argument. See todo notes in code here:
https://github.com/Lightning-AI/lightning/blob/39acb81b9bd01139f7cead31c673f25ba330f2f1/src/lightning_fabric/strategies/strategy.py#L270-L273

and  here:

https://github.com/Lightning-AI/lightning/blob/39acb81b9bd01139f7cead31c673f25ba330f2f1/src/lightning_fabric/strategies/strategy.py#L281-L282

This may not be valid for all strategies. Thus, the argument type should probably be `Optional[bool] = None`. 

### Alternatives

_No response_

### Additional context

_No response_

cc @borda @carmocca @justusschock @awaelchli

### PR DESCRIPTION
**Pull request by baskrahmer**
## What does this PR do?

Introduces a strict parameter which, when set to False, allows loading of dictionary keys even when they are not in the model state dict. Right now, strict is set to True by default to maintain backwards compatibility and to remain consistent with how PyTorch has it implemented for modules. The latter however raises the question of how we want to handle unexpected keys. Right now they are ignored, however they could also be naively added to the model_dict. This behavior could of course also be configurable.

In addition, it could be nice to log all incorrect keys in strict mode, including whether the key is missing or unexpected, just like implemented in PyTorch. Right now only the first incorrect key is logged.

Fixes #16447

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by baskrahmer** [comment]
A nice addition would be to add tests for the DeepSpeed and FSDP strategies. Not sure if this would be possible though without mocking.

File: src/lightning/fabric/strategies/deepspeed.py [hunk 1]
```diff
@@ -422,14 +422,18 @@ def save_checkpoint(
         engine.save_checkpoint(path, client_state=state, tag="checkpoint")
 
     def load_checkpoint(
-        self, path: _PATH, state: Optional[Dict[str, Union[Module, Optimizer, Any]]] = None
+        self,
+        path: _PATH,
+        state: Optional[Dict[str, Union[Module, Optimizer, Any]]] = None,
+        strict: Optional[bool] = True,
```
Comment (justusschock) [inline src/lightning/fabric/strategies/deepspeed.py:428]: I wonder if we shouldn't allow `**kwargs` here in general so that we don't need to add every argument manually?

**Review by justusschock — APPROVED** [review]
thanks for the PR @baskrahmer. Looks good to me in general. One minor point for discussion, but that's definitely optional :)

**Comment by justusschock** [comment]
> A nice addition would be to add tests for the DeepSpeed and FSDP strategies. Not sure if this would be possible though without mocking.

You should also be able to do it without mocking using something like these tests: https://github.com/Lightning-AI/lightning/blob/master/tests/tests_fabric/strategies/test_deepspeed_integration.py#L32

**Comment by baskrahmer** [comment]
> > A nice addition would be to add tests for the DeepSpeed and FSDP strategies. Not sure if this would be possible though without mocking.
> 
> You should also be able to do it without mocking using something like these tests: https://github.com/Lightning-AI/lightning/blob/master/tests/tests_fabric/strategies/test_deepspeed_integration.py#L32

Thanks for the pointer. I don't have local GPUs to run this though, so maybe someone else could pick this up.

**Review by Borda — APPROVED** [review]

File: src/lightning/fabric/strategies/deepspeed.py [hunk 2]
```diff
@@ -477,7 +481,7 @@ def load_checkpoint(
             tag="checkpoint",
             load_optimizer_states=optimzer_state_requested,
             load_lr_scheduler_states=False,
-            load_module_strict=True,  # TODO(fabric): make strict loading configurable
+            load_module_strict=strict,
```
Comment (awaelchli) [inline src/lightning/fabric/strategies/deepspeed.py:484]: This applies only to the module keys here. Below in the for-loop we also handle the client state (remaining user state dict) and strictness should also be checked there too I think. The same applies to the fsdp strategy. below. We can leave it as a follow-up for a different PR if you prefer.

File: src/lightning/fabric/strategies/strategy.py [hunk 3]
```diff
@@ -273,14 +273,19 @@ def get_optimizer_state(self, optimizer: Optimizer) -> Dict[str, Tensor]:
...
+        strict: Optional[bool] = True,
     ) -> Dict[str, Any]:
         """Load the contents from a checkpoint and restore the state of the given objects.
 
         Args:
             path: A path to where the file is located
             state: A dictionary of objects whose state will be restored in-place from the checkpoint path.
                 If no state is given, then the checkpoint will be returned in full.
+            strict (bool, optional): whether to strictly enforce that the keys
+                in `state` match the keys in the checkpoint. Default: ``True``
```
Comment (awaelchli) [inline src/lightning/fabric/strategies/strategy.py:288]:
```suggestion
            strict: whether to strictly enforce that the keys
                in `state` match the keys in the checkpoint. Default: ``True``
```

File: src/lightning/fabric/strategies/strategy.py [hunk 4]
```diff
@@ -292,19 +297,18 @@ def load_checkpoint(
             return checkpoint
 
         invalid_keys = [k for k in state if k not in checkpoint]
-        if invalid_keys:
-            # TODO(fabric): Make strict loading configurable to avoid this error if desired.
+        if strict and invalid_keys:
             raise KeyError(
-                f"The requested state contains a key '{invalid_keys[0]}' that does not exist in the loaded checkpoint."
+                f"The requested state contains a key '{invalid_keys[0]}' that does not exist in the loaded checkpoint. "
+                f"To disable strict loading, set strict to False."
```
Comment (awaelchli) [inline src/lightning/fabric/strategies/strategy.py:303]:
```suggestion
                f"To disable strict loading, set `strict=False`."
```

File: tests/tests_fabric/strategies/test_strategy.py [hunk 5]
```diff
@@ -119,4 +119,30 @@ def test_load_checkpoint_strict_loading(tmp_path):
...
+    # objects with initial state
+    saved_model = nn.Linear(2, 2)
+    saved_optimizer = torch.optim.Adam(saved_model.parameters(), lr=0.1)
+    saved_state = {"model": saved_model, "optimizer": saved_optimizer, "int": 1, "str": "test"}
+    strategy.save_checkpoint(tmp_path / "checkpoint", state=saved_state)
+
+    # same objects with different state
+    model = nn.Linear(2, 2)
+    optimizer = torch.optim.Adam(model.parameters(), lr=0.3)
+    state = {"model": model, "optimizer": optimizer, "int": 2}
```
Comment (awaelchli) [inline tests/tests_fabric/strategies/test_strategy.py:139]: This has always worked before, because we are returning the "remainder" of elements that weren't requested. What we actually need is a test case of the opposite, when state is requested to be loaded but keys are not in the state. Does that make sense? For example:

```
 state = {"model": model, "new": 111}
```

We should also check for model keys, i.e., construct a model with different submodules. 
Let me know if you have questions

File: src/lightning/fabric/strategies/fsdp.py [hunk 6]
```diff
@@ -405,7 +405,10 @@ def save_checkpoint(
             raise ValueError(f"Unknown state_dict_type: {self._state_dict_type}")
 
     def load_checkpoint(
-        self, path: _PATH, state: Optional[Dict[str, Union[Module, Optimizer, Any]]] = None
+        self,
+        path: _PATH,
+        state: Optional[Dict[str, Union[Module, Optimizer, Any]]] = None,
+        strict: Optional[bool] = None,
```
Comment (carmocca) [inline src/lightning/fabric/strategies/fsdp.py:411]: Why the Optional default?

File: src/lightning/fabric/strategies/fsdp.py [hunk 7]
```diff
@@ -457,7 +460,7 @@ def load_checkpoint(
         with state_dict_ctx:
             module_state = {module_key: module.state_dict()}
             load_state_dict(module_state, reader)
-            module.load_state_dict(module_state[module_key])
+            module.load_state_dict(module_state[module_key], strict=bool(strict))
```
Comment (carmocca) [inline src/lightning/fabric/strategies/fsdp.py:463]: Why the boolean conversion?

**Comment by awaelchli** [comment]
This looks great! Let's expose it in Fabric as well, by adding the argument to `Fabric.load()` and passing it down to the strategy call.

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
A great feature!

File: src/lightning/fabric/strategies/deepspeed.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning/fabric/strategies/deepspeed.py]: I think that we've seen some struggles in the past with mypy and inheritance with the kwargs pattern, not sure? In any case, the current code style is to write the kwargs out explicitly so I prefer to keep this consistency.
